user_groups: Add markdown rendering for group descriptions.

reharsh opened the pull request:
This PR adds support for rendered descriptions for user groups.
### Changes
1. Updated the NamedUserGroup model to include a rendered description.  
2. Refactored backend functions to handle rendered descriptions for user groups.  
3. Updated tests and documentation.  
4. Added rendered descriptions to user group settings.  
5. Group descriptions are saved and rendered on update.  
6. Rendered descriptions are shown in places like:  
- Group list Item
- Group Settings
- Typeahead  
- Popovers 
- DM

The rendered description is also sent wherever the group's description is used.

Fixes: #32559

### Screenshots

|                                    |                                     |
|-------------------------------------------|-------------------------------------------|
| User Group Settings                         | ![Changes visible in alignment and hover](https://github.com/user-attachments/assets/d8a7a000-dadb-43c7-93cc-ddfece1fa290)    |
| Direct Messages  | ![dm](https://github.com/user-attachments/assets/a107e362-481a-4bc1-807f-414319b546a8) |
| Group List Item | ![dm](https://github.com/user-attachments/assets/7183be1e-75ae-49d8-9ab6-d0aac0518abf) |
| Typeahead | ![dm](https://github.com/user-attachments/assets/3fd67bda-5bde-48e9-886f-b80c970d433c) |
| Popover | ![dm](https://github.com/user-attachments/assets/9de0c00d-cd8a-4569-bb62-396277d564fc) |


<details>
<summary>Screen Recordings</summary>


- Update/Edit user group

https://github.com/user-attachments/assets/a9f08523-c607-4b2f-b826-011c959913ca

- Create user group with link in description

https://github.com/user-attachments/assets/e926343d-9fbe-48d2-9ab2-3102ca1aecd9

</details>

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

reharsh commented:
Hey @alya , this PR is ready for review. Could you please take a look when you have a moment? Thanks!

alya commented:
The screenshots look great! Please resolve the conflicts to prepare this for review.

reharsh commented:
@alya done, please check!

alya commented:
@sahil839 could you please take a look at this one? I haven't tested.

sahil839 commented on web/src/typeahead_helper.ts:
```diff
@@ -164,10 +164,15 @@ export function rewire_render_person(value: typeof render_person): void {
         primary: user_groups.get_display_group_name(user_group.name),
-        secondary: user_group.description,
+        // eslint-disable-next-line unicorn/prefer-string-replace-all
+        secondary_html: user_group.rendered_description.replace(/^<p>|<\/p>$/g, ""),
```
Why does this need `replace` but we do not do the same for streams?

reharsh committed:
user_groups: Add markdown rendering for group descriptions.

This commit adds support for rendered descriptions for user groups.
changes:
    1. Updated the NamedUserGroup model to include a
    rendered description.
    2. Refactored backend functions to handle rendered
    descriptions for user groups.
    3. Updated tests and documentation.
    4. Added rendered descriptions to user group settings.
    5. Group descriptions are saved and rendered on update.
    6. Rendered descriptions are shown in places like:
        - Group list
        - Group Settings
        - Typeahead
        - Popovers

The rendered description is also sent wherever the group's
description is used.

Fixes: #32559

reharsh commented on zerver/lib/user_groups.py:
```diff
@@ -876,16 +879,19 @@ def create_system_user_groups_for_realm(realm: Realm) -> dict[str, NamedUserGrou
     full_members_group_info = {
         "name": SystemGroups.FULL_MEMBERS,
         "description": "Members of this organization, not including new accounts and guests",
+        "rendered_description": "<p>Members of this organization, not including new accounts and guests</p>",
```
I used `render_group_description` instead but, there was circular import issue, so I tried importing `from zerver.actions.user_groups import render_group_description` in function, it seems to work. can you please confirm if it is permitted?

reharsh commented:
Hello @sahil839, I have made the changes, sorry I was busy for a few days. can you please take a look again, when you get time. thanks!

sahil839 commented on zerver/lib/user_groups.py:
```diff
@@ -899,6 +902,13 @@ def create_system_user_groups_for_realm(realm: Realm) -> dict[str, NamedUserGrou
         everyone_on_internet_group_info,
     ]
 
+    from zerver.actions.user_groups import render_group_description
```
Can we just define `render_group_description` in this file itself, ie. `zerver/lib/user_groups.py` if that doesn't result in any import cycles?

sahil839 replied:
Ok, this seems fine for now then.

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -21256,6 +21263,10 @@ paths:
+                            rendered_description:
+                              type: string
+                              description: |
+                                The description of the user group rendered as HTML.
```
You would need to add `**Changes**` entry for all these added blocks.

sahil839 reviewed (commented):
Posted a couple of comments. Can do a complete review along with manual testing once you rebase the PR to resolve merge conflicts.

reharsh commented:
@sahil839 , made the changes, can you please take a look again. thanks!

sahil839 commented on zerver/actions/user_groups.py:
```diff
@@ -278,9 +287,13 @@ def do_update_user_group_description(
         extra_data={
             RealmAuditLog.OLD_VALUE: old_value,
             RealmAuditLog.NEW_VALUE: description,
+            "property": "description",
```
Why is this needed? We already have a different event type for description.

reharsh replied:
oh.. right, removed!

sahil839 commented on zerver/tests/test_user_groups.py:
```diff
@@ -2001,7 +2003,7 @@ def check_update_user_group(
         check_update_user_group(
             "help",
-            "Troubleshooting team",
+            "Troubleshooting team\n",
```
Would be good to add a test for checking we don't embed url previews similar to the test we have for streams.

sahil839 commented on web/src/user_groups.ts:
```diff
@@ -90,8 +100,9 @@ export function update(event: UserGroupUpdateEvent): void {
-    if (event.data.description !== undefined) {
+    if (event.data.description !== undefined && event.data.rendered_description !== undefined) {
         group.description = event.data.description;
+        group.rendered_description = event.data.rendered_description;
```
I guess `clean_up_group_description` needs to be called here as well for the new value.

sahil839 replied:
There is no need to update the `event.data.rendered_description` field here, doing `group.rendered_description = clean_up_group_description(event.data.rendered_description)` will be better.

sahil839 commented on web/src/user_group_edit.ts:
```diff
@@ -799,7 +801,12 @@ export function update_group(event: UserGroupUpdateEvent): void {
+            rendered_description: postprocess_content(group.rendered_description),
+        });
+        $edit_container.find(".group-description-wrapper").html(html);
+        $group_row.find(".description").html(html);
```
Do we need to set the value rendered by `render_user_group_description` here? This is for the left panel of groups UI and I guess that is handled by `browse_user_groups_list_item` and it handles no description case using `data-no-description` attribute.

sahil839 commented on web/src/typeahead_helper.ts:
```diff
@@ -164,10 +164,14 @@ export function rewire_render_person(value: typeof render_person): void {
     render_typeahead_item({
         primary: user_groups.get_display_group_name(user_group.name),
-        secondary: user_group.description,
+        secondary_html: user_group.rendered_description,
```
I can see that the description is not aligned correctly with the name in mention typeahead.

reharsh replied:
can you please repopulate your dev db, this is because earlier created groups don't have clean_up_descriptions

sahil839 replied:
There exists groups in production as well, so we cannot do that here. I missed this before but existing groups should be handled using migrations.

reharsh commented:
@sahil839 made the changes, PTAL.

reharsh commented:
Hello, @sahil839! I've modified migration script, please take a look when you have time.

sahil839 commented on web/src/user_group_edit.ts:
```diff
@@ -469,7 +471,10 @@ export function handle_member_edit_event(group_id: number, user_ids: number[]):
     $edit_container.find(".group-name").text(user_groups.get_display_group_name(group.name));
-    $edit_container.find(".group-description").text(group.description);
+    const html = render_user_group_description({
+        rendered_description: postprocess_content(group.rendered_description),
```
I am not sure why we do `postprocess_content` only here when updating the description in right panel and not in other cases. I guess we do the same for streams as well, but we should check why is this required only here and not in other places.

reharsh replied:
looks like `render_browse_user_group_list_item` html somehow contains postprocessed content already, like with title,target etc. and `postprocess_content` here doesn't really make any difference.

reharsh replied:
found it!
there is this handlebars helper in `template.ts` 
```javascript
Handlebars.registerHelper(
    "rendered_markdown",
    (content: string) => new Handlebars.SafeString(postprocess_content(content)),
);
```
which already does postprocessing for rendered_desc everywhere (left_sidebar, right_sidebar, typeaheads, popovers), same for stream desc, so we don't really need this `postprocess_content`.

also trying to figure out migration bit, but no luck yet, can't understand why the same thing worked for `0206_stream_rendered_description.py`:(

reharsh commented:
Hii @sahil839, I've made the changes, but getting an error while trying to migrate with new migration script.

<details>
<summary>migration file</summary>

```python

from django.db import migrations, transaction
from django.db.backends.base.schema import BaseDatabaseSchemaEditor
from django.db.migrations.state import StateApps
from django.db.models import F, Max, Min, OuterRef

def render_all_user_group_descriptions(
    apps: StateApps, schema_editor: BaseDatabaseSchemaEditor
) -> None:
    # FIXME: Application code should not be imported from migrations.
    from zerver.actions.user_groups import render_group_description

    NamedUserGroup = apps.get_model("zerver", "NamedUserGroup")
    BATCH_SIZE = 1000
    max_id = NamedUserGroup.objects.exclude(description="").aggregate(Max("id"))[
        "id__max"
    ]
    if max_id is None:
        # Do nothing if there are no user groups on the server.
        return
    
    lower_bound = NamedUserGroup.objects.exclude(description="").aggregate(Min("id"))[
        "id__min"
    ]

    while lower_bound <= max_id:
        upper_bound = lower_bound + BATCH_SIZE - 1
        print(f"Processing batch {lower_bound} to {upper_bound} for NamedUserGroup")

        with transaction.atomic():
            all_named_user_groups = NamedUserGroup.objects.exclude(
                id__gte=lower_bound, id__lt=upper_bound, description="")
            for user_group in all_named_user_groups:
                user_group.rendered_description = render_group_description(
                user_group.description, user_group.realm
                )
            NamedUserGroup.objects.bulk_update(all_named_user_groups, fields=["rendered_description"], batch_size=BATCH_SIZE)

        lower_bound += BATCH_SIZE


class Migration(migrations.Migration):

    dependencies = [
        ("zerver", "0649_namedusergroup_rendered_description"),
    ]

    operations = [
        migrations.RunPython(
            render_all_user_group_descriptions,
            reverse_code=migrations.RunPython.noop,
            elidable=True,
        ),
    ]

```
</details> 

<details><summary>Traceback</summary>

```
(zulip-py3-venv) vagrant@4b23a8f20b01:/srv/zulip$ ./manage.py migrate
Operations to perform:
  Apply all migrations: analytics, auth, confirmation, contenttypes, corporate, otp_static, otp_totp, pgroonga, phonenumber, sessions, social_django, zerver, zilencer
Running migrations:
  Applying zerver.0650_set_namedusergroup_rendered_description...Processing batch 33 to 1032 for NamedUserGroup
Traceback (most recent call last):
  File "/srv/zulip/./manage.py", line 150, in <module>
    execute_from_command_line(sys.argv)
  File "/srv/zulip/./manage.py", line 115, in execute_from_command_line
    utility.execute()
  File "/srv/zulip-venv-cache/14cbac6fefe6ac8472cebd0e64812b1d7d6157fc/zulip-py3-venv/lib/python3.10/site-packages/django/core/management/__init__.py", line 436, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/srv/zulip-venv-cache/14cbac6fefe6ac8472cebd0e64812b1d7d6157fc/zulip-py3-venv/lib/python3.10/site-packages/django/core/management/base.py", line 413, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/srv/zulip-venv-cache/14cbac6fefe6ac8472cebd0e64812b1d7d6157fc/zulip-py3-venv/lib/python3.10/site-packages/django/core/management/base.py", line 459, in execute
    output = self.handle(*args, **options)
  File "/srv/zulip-venv-cache/14cbac6fefe6ac8472cebd0e64812b1d7d6157fc/zulip-py3-venv/lib/python3.10/site-packages/django/core/management/base.py", line 107, in wrapper
    res = handle_func(*args, **kwargs)
  File "/srv/zulip-venv-cache/14cbac6fefe6ac8472cebd0e64812b1d7d6157fc/zulip-py3-venv/lib/python3.10/site-packages/django/core/management/commands/migrate.py", line 356, in handle
    post_migrate_state = executor.migrate(
  File "/srv/zulip-venv-cache/14cbac6fefe6ac8472cebd0e64812b1d7d6157fc/zulip-py3-venv/lib/python3.10/site-packages/django/db/migrations/executor.py", line 135, in migrate
    state = self._migrate_all_forwards(
  File "/srv/zulip-venv-cache/14cbac6fefe6ac8472cebd0e64812b1d7d6157fc/zulip-py3-venv/lib/python3.10/site-packages/django/db/migrations/executor.py", line 167, in _migrate_all_forwards
    state = self.apply_migration(
  File "/srv/zulip-venv-cache/14cbac6fefe6ac8472cebd0e64812b1d7d6157fc/zulip-py3-venv/lib/python3.10/site-packages/django/db/migrations/executor.py", line 252, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/srv/zulip-venv-cache/14cbac6fefe6ac8472cebd0e64812b1d7d6157fc/zulip-py3-venv/lib/python3.10/site-packages/django/db/migrations/migration.py", line 132, in apply
    operation.database_forwards(
  File "/srv/zulip-venv-cache/14cbac6fefe6ac8472cebd0e64812b1d7d6157fc/zulip-py3-venv/lib/python3.10/site-packages/django/db/migrations/operations/special.py", line 193, in database_forwards
    self.code(from_state.apps, schema_editor)
  File "/srv/zulip/zerver/migrations/0650_set_namedusergroup_rendered_description.py", line 35, in render_all_user_group_descriptions
    user_group.rendered_description = render_group_description(
  File "/srv/zulip/zerver/actions/user_groups.py", line 174, in render_group_description
    return markdown_convert(text, message_realm=realm, no_previews=True).rendered_content
  File "/srv/zulip/zerver/lib/markdown/__init__.py", line 2824, in markdown_convert
    ret = do_convert(
  File "/srv/zulip/zerver/lib/markdown/__init__.py", line 2736, in do_convert
    realm_url=message_realm.url,
AttributeError: 'Realm' object has no attribute 'url'
```
</details> 

these groups seem to have no host so I added:

```python
realm_host = getattr(user_group.realm, 'host', None)
if not realm_host:
        print(f"Skipping user group {user_group.name}: realm host is missing")
        continue
```

<details><summary>output is...</summary>
Skipping user group role:nobody: realm host is missing
Skipping user group role:owners: realm host is missing
Skipping user group role:administrators: realm host is missing
Skipping user group role:moderators: realm host is missing
Skipping user group role:fullmembers: realm host is missing
Skipping user group role:members: realm host is missing
Skipping user group role:everyone: realm host is missing
Skipping user group role:internet: realm host is missing
Skipping user group role:nobody: realm host is missing
Skipping user group role:owners: realm host is missing
Skipping user group role:administrators: realm host is missing
Skipping user group role:moderators: realm host is missing
Skipping user group role:fullmembers: realm host is missing
Skipping user group role:members: realm host is missing
Skipping user group role:everyone: realm host is missing
Skipping user group role:internet: realm host is missing
Skipping user group hamletcharacters: realm host is missing
Skipping user group role:nobody: realm host is missing
Skipping user group role:owners: realm host is missing
Skipping user group role:administrators: realm host is missing
Skipping user group role:moderators: realm host is missing
Skipping user group role:fullmembers: realm host is missing
Skipping user group role:members: realm host is missing
Skipping user group role:everyone: realm host is missing
Skipping user group role:internet: realm host is missing
Skipping user group new bolds: realm host is missing
Skipping user group new italics: realm host is missing
Skipping user group this is link: realm host is missing
</details> 

 If you have time can you please look into it.

sahil839 reviewed (commented):
Posted a couple of comments.

Regarding migration, I tried running `./manage.py migrate` with the migration file that you have in this PR without any changes and it gives the same error that you mentioned. I think it is because `uri` is a property on `Realm` and not a DB field, but am not sure.

reharsh commented:
Thank you, @sahil839, for your time on this!

I just figured out that running all migrations together (`./tools/provision --force`) resolves the issue, likely due to a model state inconsistency with the custom `RunPython` code was causing it. It seems consistent with other files too, I think this is expected, given migrations are applied sequentially. Let me know if you have any other thoughts.
again thank you so much -- I've made the changes—please review them whenever you have time.

sahil839 commented:
You can do provisioning and rebuild the DB in development environment, but that cannot be done in production. Same migration file for stream description works, so am not sure how to fix this.

reharsh commented:
Ok... then I'll continue to look for a fix,
but while running `stream_rendered_description` by `./manage.py migrate zerver 0206_stream_rendered_description` or any other old migrations it gives me following error. I can't run those too... like you said it works for you. any insights?

<details><summary>traceback</summary>

```
(zulip-py3-venv) vagrant@4b23a8f20b01:/srv/zulip$ ./manage.py migrate zerver 0206_stream_rendered_description
Operations to perform:
  Target specific migration: 0206_stream_rendered_description, from zerver
Traceback (most recent call last):
  File "/srv/zulip/./manage.py", line 150, in <module>
    execute_from_command_line(sys.argv)
  File "/srv/zulip/./manage.py", line 115, in execute_from_command_line
    utility.execute()
  File "/srv/zulip-venv-cache/14cbac6fefe6ac8472cebd0e64812b1d7d6157fc/zulip-py3-venv/lib/python3.10/site-packages/django/core/management/__init__.py", line 436, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/srv/zulip-venv-cache/14cbac6fefe6ac8472cebd0e64812b1d7d6157fc/zulip-py3-venv/lib/python3.10/site-packages/django/core/management/base.py", line 413, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/srv/zulip-venv-cache/14cbac6fefe6ac8472cebd0e64812b1d7d6157fc/zulip-py3-venv/lib/python3.10/site-packages/django/core/management/base.py", line 459, in execute
    output = self.handle(*args, **options)
  File "/srv/zulip-venv-cache/14cbac6fefe6ac8472cebd0e64812b1d7d6157fc/zulip-py3-venv/lib/python3.10/site-packages/django/core/management/base.py", line 107, in wrapper
    res = handle_func(*args, **kwargs)
  File "/srv/zulip-venv-cache/14cbac6fefe6ac8472cebd0e64812b1d7d6157fc/zulip-py3-venv/lib/python3.10/site-packages/django/core/management/commands/migrate.py", line 301, in handle
    pre_migrate_state = executor._create_project_state(with_applied_migrations=True)
  File "/srv/zulip-venv-cache/14cbac6fefe6ac8472cebd0e64812b1d7d6157fc/zulip-py3-venv/lib/python3.10/site-packages/django/db/migrations/executor.py", line 91, in _create_project_state
    migration.mutate_state(state, preserve=False)
  File "/srv/zulip-venv-cache/14cbac6fefe6ac8472cebd0e64812b1d7d6157fc/zulip-py3-venv/lib/python3.10/site-packages/django/db/migrations/migration.py", line 91, in mutate_state
    operation.state_forwards(self.app_label, new_state)
  File "/srv/zulip-venv-cache/14cbac6fefe6ac8472cebd0e64812b1d7d6157fc/zulip-py3-venv/lib/python3.10/site-packages/django/db/migrations/operations/fields.py", line 165, in state_forwards
    state.remove_field(app_label, self.model_name_lower, self.name)
  File "/srv/zulip-venv-cache/14cbac6fefe6ac8472cebd0e64812b1d7d6157fc/zulip-py3-venv/lib/python3.10/site-packages/django/db/migrations/state.py", line 258, in remove_field
    old_field = model_state.fields.pop(name)
KeyError: 'create_public_stream_policy'
```
</details>

sahil839 commented:
I meant I have not seen any previous discussion of any such failure when the stream rendered description work was done. The way you are trying to run the single migration run will expectedly not work because the Database structure has change a lot since that migration.

I think would be better to ask for suggestions on CZO if you cannot get a fix.

timabbott commented on zerver/migrations/0698_namedusergroup_rendered_description.py:
```diff
@@ -0,0 +1,17 @@
+        migrations.AddField(
+            model_name="namedusergroup",
+            name="rendered_description",
+            field=models.TextField(default=""),
```
Provide a `db_default` for this -- otherwise, it'll be an expensive migration to apply in production.

Also, is `""` what you get when running the Markdown processor on the empty string? We likely want to match that.

timabbott commented on tools/semgrep-py.yml:
```diff
@@ -84,6 +84,7 @@ rules:
         - zerver/migrations/0443_userpresence_new_table_schema.py
         - zerver/migrations/0493_rename_userhotspot_to_onboardingstep.py
         - pgroonga/migrations/0002_html_escape_subject.py
+        - zerver/migrations/0650_set_namedusergroup_rendered_description.py
```
This is out of date.

timabbott commented on version.py:
```diff
@@ -34,7 +34,7 @@
 # entries in the endpoint's documentation in `zulip.yaml`.
 
-API_FEATURE_LEVEL = 379
+API_FEATURE_LEVEL = 380  # Last bumped for adding rendered_description to user groups.
```
You can use `tools/create-api-changelog` to avoid busywork updating this now.

timabbott commented on zerver/migrations/0699_set_namedusergroup_rendered_description.py:
```diff
@@ -0,0 +1,56 @@
+            for user_group in all_named_user_groups:
+                user_group.rendered_description = render_group_description(
+                    user_group.description, user_group.realm
+                )
```
I don't think it's likely safe to run the Markdown processor inside a transaction.

laurynmm commented:
Closing this in favor of #38473, which is an updated/rebased version of these changes and moving the completion candidate label to that PR.
